src/databricks/labs/ucx/migration/azure_credentials.py

```diff
@@ -0,0 +1,170 @@
+        self._ws = ws
+        self._azure_resource_permissions = azure_resource_permissions
+        self._azure_sp_crawler = azure_sp_crawler
+        self._action_plan = 'service_principals_for_storage_credentials.csv'
+
+
+    @classmethod
+    def for_cli(cls, ws: WorkspaceClient, customized_csv: str, replace_with_ac: bool, product='ucx'):
+        installation = Installation.current(ws, product)
+        config = installation.load(WorkspaceConfig)
+        sql_backend = StatementExecutionBackend(ws, config.warehouse_id)
+        azurerm = AzureResources(ws)
+        locations = ExternalLocations(ws, sql_backend, config.inventory_database)
+
+        azure_resource_permissions = AzureResourcePermissions(installation, ws, azurerm, locations)
+        azure_sp_crawler = AzureServicePrincipalCrawler(ws, sql_backend, config.inventory_database)
+
+        return cls(installation, ws, azure_resource_permissions, azure_sp_crawler, customized_csv, replace_with_ac)
+
+
+    def _list_storage_credentials(self):
+        # list existed storage credentials that is using service principal, capture the service principal's application_id
+        return {}
+
+
+    def _check_sp_in_storage_credentials(self, sp_list, sc_set):
+        # if sp is already used, take it off from the sp_list
+        return list()
+
+
+    def _fetch_client_secret(self, sp_list: list[StoragePermissionMapping]) -> list[ServicePrincipalMigrationInfo]:
+        # check AzureServicePrincipalInfo from AzureServicePrincipalCrawler, if AzureServicePrincipalInfo
+        # has secret_scope and secret_key not empty, fetch the client_secret and put it to the client_secret field
+        #
+        # The input StoragePermissionMapping may have managed identity mixed in, we will ignore them for now, as
+        # they won't have any client_secret, we will process managed identity in the future.
+
+        # fetch client_secrets of crawled service principal, if any
+        azure_sp_info_with_client_secret = {}
+        azure_sp_infos = self._azure_sp_crawler.snapshot()
+
+        for azure_sp_info in azure_sp_infos:
+            if azure_sp_info.secret_scope is None:
+                continue
+            if azure_sp_info.secret_key is None:
+                continue
+
+            try:
+                secret_response = self._ws.secrets.get_secret(azure_sp_info.secret_scope, azure_sp_info.secret_key)
+            except Unauthenticated:
+                logger.info(f"User is unauthenticated to fetch secret value. Cannot fetch the service principal "
+                            f"client_secret for {azure_sp_info.application_id}. Will not reuse this client_secret")
+                continue
+            except PermissionDenied:
+                logger.info(f"User does not have permission to read secret value for {azure_sp_info.secret_scope}.{azure_sp_info.secret_key}. "
+                            f"Cannot fetch the service principal client_secret for {azure_sp_info.application_id}. "
```

Comment:
Also, why do we log this particular case? I think we should rethrow, as UCX is run as an workspace admin